micro-rdk/src/common/app_client.rs

```diff
@@ -285,10 +315,58 @@ impl AppClient {
 
         Ok(())
     }
+
+    /// Obtains the Duration for which we should wait before next
+    /// checking for a restart. If no Duration is returned, then the
+    /// app has signaled that we should restart now.
+    pub async fn check_for_restart(&mut self) -> Result<Option<Duration>, AppClientError> {
+        let req = NeedsRestartRequest {
+            id: self.config.robot_id.clone(),
+        };
+        let body = encode_request(req)?;
+        let r = self
+            .grpc_client
+            .build_request(
+                "/viam.app.v1.RobotService/NeedsRestart",
+                Some(&self.jwt),
+                "",
+                BodyExt::boxed(Full::new(body).map_err(|never| match never {})),
+            )
+            .map_err(AppClientError::AppGrpcClientError)?;
+        let (mut response, headers_) = self.grpc_client.send_request(r).await?;
+        let response = NeedsRestartResponse::decode(response.split_off(5))?;
+        if response.must_restart {
+            Ok(None)
+        } else {
+            const MIN_RESTART_DURATION: Duration = Duration::from_secs(1);
+            const DEFAULT_RESTART_DURATION: Duration = Duration::from_secs(5);
+            Ok(response
+                .restart_check_interval
+                .map_or_else(
+                    || DEFAULT_RESTART_DURATION,
+                    |d| {
+                        Duration::try_from(d).map_or_else(
+                            |e| DEFAULT_RESTART_DURATION,
+                            |d| d.max(MIN_RESTART_DURATION),
```

Comment:
This can probably be simplified a bit so that each constant only needs to be used once.